locator/abstract_replication_strategy.cc

```diff
@@ -169,6 +169,12 @@ effective_replication_map::get_ranges(inet_address ep) const {
     });
 }
 
+future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(std::optional<endpoint_set>& eps, const token& search_token, const token_metadata& tm) const {
+    if (!eps || !has_static_natural_endpoints()) {
+        eps = co_await calculate_natural_endpoints(search_token, tm);
+    }
```

Comment:
For Everywhere, don't the endpoints change when nodes are added or removed?

Reply:
Yes, but the topology never changes in the context of the eps, it holds an immutable token_metadata (that encapsulates the topology)

Reply:
But I see you point that this happens outside the scope of this function

Reply:
Agreed.  I'm about to revise it using a safe approach that detects if the token_metadata changed using its version. Although it shouldn't none of the current uses in this series, the method is exposed as public and external users may not respect that convention.

```diff
diff --git a/locator/abstract_replication_strategy.cc b/locator/abstract_replication_strategy.cc
index 4ffee1b10e..f49b69985c 100644
--- a/locator/abstract_replication_strategy.cc
+++ b/locator/abstract_replication_strategy.cc
@@ -169,9 +169,12 @@ effective_replication_map::get_ranges(inet_address ep) const {
     });
 }
 
-future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(std::optional<endpoint_set>& eps, const token& search_token, const token_metadata& tm) const {
-    if (!eps || !has_static_natural_endpoints()) {
-        eps = co_await calculate_natural_endpoints(search_token, tm);
+future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(cached_endpoint_set& cached_eps, const token& search_token, const token_metadata& tm) const {
+    auto cur_ring_version = tm.get_ring_version();
+    assert(cur_ring_version >= 0);
+    if (!has_static_natural_endpoints() || cached_eps.ring_version != cur_ring_version) {
+        cached_eps.eps = co_await calculate_natural_endpoints(search_token, tm);
+        cached_eps.ring_version = cur_ring_version;
     }
 }
 
@@ -184,7 +187,7 @@ abstract_replication_strategy::get_ranges(inet_address ep, token_metadata_ptr tm
     if (sorted_tokens.empty()) {
         on_internal_error(rslogger, "Token metadata is empty");
     }
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
     auto prev_tok = sorted_tokens.back();
     for (auto tok : sorted_tokens) {
         co_await maybe_calculate_natural_endpoints(eps, tok, tm);
@@ -223,7 +226,7 @@ effective_replication_map::get_primary_ranges_within_dc(inet_address ep) const {
 future<std::unordered_multimap<inet_address, dht::token_range>>
 abstract_replication_strategy::get_address_ranges(const token_metadata& tm) const {
     std::unordered_multimap<inet_address, dht::token_range> ret;
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
 
     for (auto& t : tm.sorted_tokens()) {
         dht::token_range_vector r = tm.get_primary_ranges_for(t);
@@ -244,7 +247,7 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
     if (!tm.is_member(endpoint)) {
         co_return ret;
     }
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
     bool is_everywhere_topology = get_type() == replication_strategy_type::everywhere_topology;
     for (auto& t : tm.sorted_tokens()) {
         // This is a fast path for everywhere_topology to avoid calculating
@@ -265,7 +268,7 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
                 ret.emplace(endpoint, rng);
             }
         } else {
-            rslogger.debug("token={} natural_endpoints={}: endpoint={} not found", t, eps, endpoint);
+            rslogger.debug("token={} natural_endpoints={}: endpoint={} not found", t, *eps, endpoint);
         }
     }
     co_return ret;
@@ -288,7 +291,7 @@ effective_replication_map::get_range_addresses() const {
 future<std::unordered_map<dht::token_range, inet_address_vector_replica_set>>
 abstract_replication_strategy::get_range_addresses(const token_metadata& tm) const {
     std::unordered_map<dht::token_range, inet_address_vector_replica_set> ret;
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
 
     for (auto& t : tm.sorted_tokens()) {
         dht::token_range_vector ranges = tm.get_primary_ranges_for(t);
@@ -311,7 +314,7 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
     token_metadata temp;
     temp = co_await tmptr->clone_only_token_map();
     co_await temp.update_normal_tokens(pending_tokens, pending_address);
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
 
     for (const auto& t : temp.sorted_tokens()) {
         co_await maybe_calculate_natural_endpoints(eps, t, temp);
@@ -327,7 +330,7 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
 
 future<mutable_effective_replication_map_ptr> calculate_effective_replication_map(abstract_replication_strategy::ptr_type rs, token_metadata_ptr tmptr) {
     replication_map replication_map;
-    std::optional<endpoint_set> eps;
+    abstract_replication_strategy::cached_endpoint_set eps;
 
     for (const auto &t : tmptr->sorted_tokens()) {
         co_await rs->maybe_calculate_natural_endpoints(eps, t, *tmptr);
diff --git a/locator/abstract_replication_strategy.hh b/locator/abstract_replication_strategy.hh
index f37ec38698..982915f067 100644
--- a/locator/abstract_replication_strategy.hh
+++ b/locator/abstract_replication_strategy.hh
@@ -82,7 +82,29 @@ class abstract_replication_strategy {
 
     virtual bool has_static_natural_endpoints() const noexcept { return false; }
 
-    future<> maybe_calculate_natural_endpoints(std::optional<endpoint_set>& eps, const token& search_token, const token_metadata& tm) const;
+    // Avoids redundant calculations
+    struct cached_endpoint_set {
+        endpoint_set eps;
+        long ring_version = -1;
+
+        const endpoint_set& operator*() const noexcept {
+            return eps;
+        }
+
+        endpoint_set& operator*() noexcept {
+            return eps;
+        }
+
+        const endpoint_set* operator->() const noexcept {
+            return &eps;
+        }
+
+        endpoint_set* operator->() noexcept {
+            return &eps;
+        }
+    };
+
+    future<> maybe_calculate_natural_endpoints(cached_endpoint_set& eps, const token& search_token, const token_metadata& tm) const;
 
     // The returned vector has size O(number of normal token owners), which is O(number of nodes in the cluster).
     // Note: it is not guaranteed that the function will actually yield. If the complexity of a particular implementation
```